tests/test_atom_from_annotations.py

```diff
@@ -0,0 +1,202 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) 2021, Nucleic Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ------------------------------------------------------------------------------
+"""Test defining an atom class using typing annotations.
+
+"""
+import sys
+from typing import (
+    Any,
+    Callable as TCallable,
+    ClassVar,
+    Dict as TDict,
+    Iterable,
+    List as TList,
+    Optional,
+    Set as TSet,
+)
+
+import pytest
+
+from atom.api import (
+    Atom,
+    Bool,
+    Bytes,
+    Callable,
+    Dict,
+    Float,
+    Instance,
+    Int,
+    List,
+    Member,
+    Set,
+    Str,
+    Value,
+)
+from atom.atom import set_default
+
+
+def test_ignore_annotations():
+    class A(Atom):
+        a: int
+
+    assert not hasattr(A, "a")
+
+
+def test_reject_str_annotations():
+    with pytest.raises(TypeError):
+
+        class A(Atom, use_annotations=True):
+            a: "int"
+
+
+def test_ignore_class_var():
+    class A(Atom, use_annotations=True):
+        a: ClassVar[int] = 1
+
+    assert not isinstance(A.a, Member)
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason="Subscription of Members requires Python 3.9+"
+)
+def test_ignore_annotated_member():
+    class A(Atom, use_annotations=True):
+        a: List[int] = List(int, default=[1, 2, 3])
```

Comment:
I'm not sure I understand the intent of this code.  I thought this would be a syntax error.

Reply:
The point is that if the user also defines, for example, a post-validate he may need to narrow down further the type.

What I am trying to avoid is complete garbage annotation like a: dict = Dict() without blocking possibly valid usages.

Reply:
We cannot disallow it in general. Making atom members understandable to type checkers mean you can use a member as a type annotation. One thing I want to allow at least is:

````python

class A(Atom):
 
    o: Member[int, str] = Value()

    def _validate_o(self, old, new):
        assert isinstance(new, str)
        return o

    def _post_getattr_o(self, value):
        return int(value)

```

Reply:
Oh, neat.  I didn't know you had implemented that.  My inclination would have been to do something like:

```python
class A(Atom):
    o: Union[int, str]
```

which I would interpret to mean `o = Instance(kind=(int, str))`, and the same validators/getters could be applied if you wanted.

Reply:
Also, I think `Member[int, str] = Value()` has similar issues to combining annotations with member declarations.

Reply:
Actually Member[int, str] means you can set a str but get an int which is different.

Reply:
I find it less problematic than allowing `a: int = Int()` which is always straight up garbage. Here we simply allow user to provide finer grained information to the static type checker that it would otherwise be unable to infer. I find it extremely annoying when atom enforce that typing is correct at runtime but a static type checker fails to understand it. The only case I consider it acceptable is when we are doing value checking in atom.

Reply:
However one thing we can and probably should disallow is a bare member annotation since this is meaningless.

```python
class A(Atom):
    a : Int

```

Reply:
You can have a look at the .pyi for CAtom. Member is defined as a descriptor generic over 2 types (https://github.com/nucleic/atom/blob/main/atom/catom.pyi#L50), the type returned upon access and the type accepted when setting (https://github.com/nucleic/atom/blob/main/atom/catom.pyi#L67 and https://github.com/nucleic/atom/blob/main/atom/catom.pyi#L68).

This is perfectly valid from the point of view of a type checker and does not involve mypy plugin so it should work just as well with pyright.

Reply:
Thanks for pointing that out.  Makes sense now.  Just trying to generalize the usage in my mind.  Why not use the member sublass as the annotation (e.g. `Value[int, str]` or `Instance[int, str]`)?  (I see that instance.pyi declare a single get/set type, but not sure of the underlying restriction).  Or, are you restricting the usage of members as type hints to just the `Member` class.

Reply:
I completely agree. To summarize we have:
```python
# Valid
a: int
a = Int()
a: "Int" = Int()  # -> probably the one we would like to exclude but cannot
a: Int = Int()
# Invalid
a: Int
a: int = Int()
```

Reply:
For reference/completeness, atom throws the following exceptions for the others:
```python
a: Int          # ValueError: Member subclasses cannot be used as annotations without specifying a default value for the attribute.
a: int = Int()  # TypeError: Field 'a' of 'A' is assigned a Member-like value but its annotation is not Member compatible
```

Reply:
I've also verified that mypy flags invalid assignments to `a` for the first two cases (e.g. `a.a = "a"` gives the error ` Incompatible types in assignment (expression has type "str", variable has type "int")`

```
a: int
a = Int()
```

Reply:
So the Int stuff is kinda weird. Because atom has a non strict mode accepting float  Int is actually generic over the type that can be set. So my example in addition of being simplistic is actually wrong. I will try to come with something better and add a mypy test.